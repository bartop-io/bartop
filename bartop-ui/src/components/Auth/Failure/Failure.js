import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';

import strings from '../../../strings';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100%;
`;

export const FailureMessage = styled.p`
  /* stylelint-disable-line block-no-empty */
`;

export const Failure = ({ error }) => (
  <Wrapper>
    <FailureMessage>{`${strings.auth.failure} ${
      error ? error.message : ''
    }`}</FailureMessage>
  </Wrapper>
);

Failure.propTypes = {
  error: PropTypes.object
};

const mapStateToProps = state => ({
  error: state.authentication.status.error
});

export default connect(mapStateToProps)(Failure);
