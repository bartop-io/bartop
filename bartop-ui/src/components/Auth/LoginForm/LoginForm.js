import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';

const LoginForm = ({ history, sendCode, prefillEmail }) => (
  <Formik
    initialValues={{
      email: prefillEmail || ''
    }}
    validate={({ email }) => {
      const errors = {};
      if (!email) {
        errors.email = 'Required';
      } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(email)) {
        errors.email = 'Invalid email address';
      }
      return errors;
    }}
    onSubmit={async ({ email }, { setSubmitting, setFieldError }) => {
      try {
        await sendCode(email);
        setSubmitting(false);
        history.replace({ pathname: '/auth/verify', state: { email } });
      } catch (err) {
        setSubmitting(false);
        setFieldError('email', err.message);
      }
    }}
    render={({
      values,
      errors,
      touched,
      handleChange,
      handleBlur,
      handleSubmit,
      isSubmitting
    }) => (
      <form onSubmit={handleSubmit}>
        <p>Enter your email to sign in or create an account</p>
        <input
          type="email"
          name="email"
          placeholder="bilbo@bartop.io"
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.email}
        />
        {touched.email && errors.email && <div>{errors.email}</div>}
        <button type="submit" disabled={isSubmitting}>
          Submit
        </button>
      </form>
    )}
  />
);

LoginForm.propTypes = {
  history: PropTypes.object.isRequired,
  sendCode: PropTypes.func.isRequired,
  prefillEmail: PropTypes.string
};

export default LoginForm;
