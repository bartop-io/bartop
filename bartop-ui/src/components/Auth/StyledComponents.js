import { Form } from 'formik';
import styled from 'styled-components';

export const StyledForm = styled(Form)`
  width: 100%;
  max-width: 600px;
  height: 100%;
  max-height: 600px;
  margin: 0 auto;
  padding: 10px;
`;

export const PromptContainer = styled.div`
  height: 40%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
`;

export const Prompt = styled.p`
  font-size: medium;
`;

export const InputContainer = styled.div`
  height: 40%;
  display: flex;
  align-items: center;
`;

export const SubmitContainer = styled.div`
  height: 20%;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const SubmitButton = styled.button`
  &:hover:enabled {
    cursor: pointer;
  }
`;
