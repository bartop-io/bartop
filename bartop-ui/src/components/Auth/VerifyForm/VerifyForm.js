import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';

const VerifyForm = ({ history, email, verifyCode }) => (
  <Formik
    initialValues={{
      verificationCode: ''
    }}
    validate={({ verificationCode }) => {
      const errors = {};
      if (!verificationCode) {
        errors.verificationCode = 'Required';
      } else if (verificationCode.length !== 6) {
        errors.verificationCode = 'Code must be six digits';
      }
      return errors;
    }}
    onSubmit={async (
      { verificationCode },
      { setSubmitting, setFieldError }
    ) => {
      try {
        await verifyCode(email, verificationCode);
        setSubmitting(false);
        history.replace({
          pathname: '/auth/callback'
        });
      } catch (err) {
        setSubmitting(false);
        setFieldError('verificationCode', err.message);
      }
    }}
    render={({
      values,
      errors,
      touched,
      handleChange,
      handleBlur,
      handleSubmit,
      isSubmitting
    }) => (
      <form onSubmit={handleSubmit}>
        <p>An email with the code has been sent to {email}</p>
        <input
          type="text"
          name="verificationCode"
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.verificationCode}
        />
        {touched.verificationCode &&
          errors.verificationCode && <div>{errors.verificationCode}</div>}
        <button type="submit" disabled={isSubmitting}>
          Submit
        </button>
        <div
          onClick={() =>
            history.replace({ pathname: '/auth/login', state: { email } })
          }
        >
          Didn&#39;t receive the code?
        </div>
      </form>
    )}
  />
);

VerifyForm.propTypes = {
  history: PropTypes.object.isRequired,
  email: PropTypes.string.isRequired,
  verifyCode: PropTypes.func.isRequired
};

export default VerifyForm;
