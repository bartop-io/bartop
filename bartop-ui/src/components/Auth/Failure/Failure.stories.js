import React from 'react';

import { storiesOf } from '@storybook/react';

import { Failure } from './Failure';

storiesOf('Failure', module)
  .add('default', () => <Failure />)
  .add('with an error message', () => (
    <Failure error={new Error('Something went horribly wrong!')} />
  ));
