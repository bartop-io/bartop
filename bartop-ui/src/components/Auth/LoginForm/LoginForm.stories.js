import React from 'react';

import { storiesOf } from '@storybook/react';

import LoginForm from './LoginForm';
import { noop } from '../../../test-helpers/utils';

const requiredProps = {
  history: {},
  sendCode: noop
};

storiesOf('LoginForm', module)
  .add('with required props', () => <LoginForm {...requiredProps} />)
  .add('with a prefillEmail', () => (
    <LoginForm {...requiredProps} prefillEmail="john@smith.com" />
  ));
