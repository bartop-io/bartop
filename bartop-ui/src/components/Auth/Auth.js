import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Switch, Route } from 'react-router-dom';

import auth from '../../singletons/authentication';
import LoginForm from './LoginForm/LoginForm';
import VerifyForm from './VerifyForm/VerifyForm';
import Callback from './Callback/Callback';
import Failure from './Failure/Failure';
import { actions } from '../../ducks/authentication/authentication';
import NotFound from '../NotFound/NotFound';

export class Auth extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      sendCodeError: null,
      verificationCode: '',
      verifyCodeError: null
    };
    this.setEmail = this.setEmail.bind(this);
    this.setCode = this.setCode.bind(this);
    this.sendCode = this.sendCode.bind(this);
    this.verifyCode = this.verifyCode.bind(this);
  }

  setEmail(email) {
    this.setState({
      email
    });
  }

  setCode(verificationCode) {
    this.setState({
      verificationCode
    });
  }

  sendCode() {
    const { history } = this.props;
    const { email } = this.state;
    auth.passwordlessStart(
      {
        connection: 'email',
        send: 'code',
        email: this.state.email
      },
      (err, res) => {
        if (err) {
          console.error(err);
          this.setState({
            sendCodeError: err
          });
        } else {
          history.replace({
            pathname: '/auth/verify',
            state: {
              email
            }
          });
        }
      }
    );
  }

  verifyCode() {
    auth.passwordlessLogin(
      {
        connection: 'email',
        email: this.state.email,
        verificationCode: this.state.verificationCode
      },
      (err, res) => {
        if (err) {
          this.setState({
            verifyCodeError: err
          });
        } else {
          console.log('res is', res);
        }
      }
    );
  }

  render() {
    const { match, handleAuthentication } = this.props;
    const {
      email,
      sendCodeError,
      verificationCode,
      verifyCodeError
    } = this.state;

    return (
      <Switch>
        <Route
          exact
          path={`${match.url}/login`}
          render={() => (
            <LoginForm
              email={email}
              setEmail={this.setEmail}
              sendCode={this.sendCode}
              sendCodeError={sendCodeError}
            />
          )}
        />
        <Route
          exact
          path={`${match.url}/verify`}
          render={({ location }) => (
            <VerifyForm
              email={email}
              verificationCode={verificationCode}
              setCode={this.setCode}
              verifyCode={this.verifyCode}
              verifyCodeError={verifyCodeError}
            />
          )}
        />
        <Route
          exact
          path={`${match.url}/callback`}
          render={() => {
            handleAuthentication();
            return <Callback />;
          }}
        />
        <Route exact path={`${match.url}/failure`} component={Failure} />
        <Route component={NotFound} />
      </Switch>
    );
  }
}

Auth.propTypes = {
  history: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  handleAuthentication: PropTypes.func.isRequired
};

const mapStateToProps = state => ({});

const mapDispatchToProps = {
  handleAuthentication: actions.handleAuthentication
};

export default connect(mapStateToProps, mapDispatchToProps)(Auth);
