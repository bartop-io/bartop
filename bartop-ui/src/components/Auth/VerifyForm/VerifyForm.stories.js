import React from 'react';

import { storiesOf } from '@storybook/react';

import VerifyForm from './VerifyForm';
import { noop } from '../../../test-helpers/utils';

const requiredProps = {
  history: {},
  email: 'john@smith.com',
  verifyCode: noop
};

storiesOf('VerifyForm', module).add('with required props', () => (
  <VerifyForm {...requiredProps} />
));
