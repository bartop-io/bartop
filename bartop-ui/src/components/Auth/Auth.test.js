import React from 'react';
import { shallow, mount } from 'enzyme';
import { StaticRouter, Route } from 'react-router-dom';
import 'jest-styled-components';

import { Auth } from './Auth';
import LoginForm from './LoginForm/LoginForm';
import VerifyForm from './VerifyForm/VerifyForm';
import Callback from './Callback/Callback';
import NotFound from '../NotFound/NotFound';
import Failure from './Failure/Failure';

jest.mock('./Failure/Failure');

const requiredProps = {
  match: {},
  handleAuthentication: jest.fn(),
  auth: {
    passwordlessStart: jest.fn(),
    passwordlessLogin: jest.fn()
  }
};

const mountRouterAtLocation = location =>
  mount(
    <StaticRouter context={{}} location={location}>
      <Route
        path="/auth"
        render={({ match }) => <Auth {...requiredProps} match={match} />}
      />
    </StaticRouter>
  );

it('matches the snapshot', () => {
  const wrapper = shallow(
    <StaticRouter context={{}}>
      <Auth {...requiredProps} />
    </StaticRouter>
  );
  expect(wrapper).toMatchSnapshot();
});

it('if path is /auth/login, renders LoginForm', () => {
  const wrapper = mountRouterAtLocation('/auth/login');
  expect(wrapper.find(LoginForm).exists()).toBe(true);
});

// TODO - determine why setting location state is only not working in tests
xit('if path is /auth/login and email is in location state, renders LoginForm with prefillEmail prop', () => {
  const wrapper = mountRouterAtLocation({
    pathname: '/auth/login',
    state: { email: 'bilbo@bartop.io' }
  });
  expect(wrapper.find(LoginForm).exists()).toBe(true);
  expect(wrapper.find(LoginForm).prop('prefillEmail')).toEqual(
    'bilbo@bartop.io'
  );
});

// TODO - determine why setting location state is only not working in tests
xit('if path is /auth/verify and email is in location state, renders VerifyForm', () => {
  const wrapper = mountRouterAtLocation({
    pathname: '/auth/verify',
    state: { email: 'bilbo@bartop.io' }
  });
  expect(wrapper.find(VerifyForm).exists()).toBe(true);
  expect(wrapper.find(VerifyForm).prop('prefillEmail')).toEqual(
    'bilbo@bartop.io'
  );
});

it('if path is /auth/verify and email is not in location state, renders NotFound', () => {
  const wrapper = mountRouterAtLocation('/auth/verify');
  expect(wrapper.find(NotFound).exists()).toBe(true);
});

it('if path is /auth/callback, renders Callback and calls handleAuthentication()', () => {
  const wrapper = mountRouterAtLocation('/auth/callback');
  expect(wrapper.find(Callback).exists()).toBe(true);
  expect(requiredProps.handleAuthentication).toBeCalled();
});

it('if path is /auth/failure, renders Failure', () => {
  const wrapper = mountRouterAtLocation('/auth/failure');
  expect(wrapper.find(Failure).exists()).toBe(true);
});

it('if path is anything else, renders NotFound', () => {
  let wrapper = mountRouterAtLocation('/auth');
  expect(wrapper.find(NotFound).exists()).toBe(true);

  wrapper = mountRouterAtLocation('/auth/no-jawn-here');
  expect(wrapper.find(NotFound).exists()).toBe(true);
});
