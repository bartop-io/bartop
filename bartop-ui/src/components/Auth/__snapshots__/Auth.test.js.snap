// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Auth
    auth={
      Object {
        "passwordlessLogin": [Function],
        "passwordlessStart": [Function],
      }
    }
    handleAuthentication={[Function]}
    match={Object {}}
  />
</Router>
`;
